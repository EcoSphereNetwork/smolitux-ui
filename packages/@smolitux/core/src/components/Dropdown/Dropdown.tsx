// packages/@smolitux/core/src/components/Dropdown/Dropdown.tsx
import React, { useState, useRef, useEffect, createContext, useContext, useCallback } from 'react';
import ReactDOM from 'react-dom';

// Dropdown-Context für Zustände und Funktionen
type DropdownContextType = {
  isOpen: boolean;
  setIsOpen: (isOpen: boolean) => void;
  activeItemIndex: number | null;
  setActiveItemIndex: (index: number | null) => void;
  registerItem: (id: string) => number;
  triggerRef: React.RefObject<HTMLElement>;
  dropdownId: string;
  onSelect?: (value: string) => void;
  closeDropdown: () => void;
};

const DropdownContext = createContext<DropdownContextType | undefined>(undefined);

// Hook für Dropdown-Context
export const useDropdownContext = () => {
  const context = useContext(DropdownContext);
  if (context === undefined) {
    throw new Error('useDropdownContext must be used within a Dropdown component');
  }
  return context;
};

export interface DropdownProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Kinder-Elemente (DropdownToggle, DropdownMenu) */
  children: React.ReactNode;
  /** Ist das Dropdown geöffnet (kontrollierter Modus) */
  isOpen?: boolean;
  /** Callback beim Öffnen/Schließen */
  onOpenChange?: (isOpen: boolean) => void;
  /** Callback beim Öffnen */
  onOpen?: () => void;
  /** Callback beim Schließen */
  onClose?: () => void;
  /** Callback bei Item-Auswahl */
  onSelect?: (value: string) => void;
  /** Position des Dropdowns relativ zum Trigger */
  placement?: 'bottom' | 'top' | 'left' | 'right';
  /** Größe des Dropdowns */
  size?: 'sm' | 'md' | 'lg';
  /** Ist das Dropdown deaktiviert? */
  isDisabled?: boolean;
  /** Bei Klick außerhalb schließen */
  closeOnClickOutside?: boolean;
  /** Bei ESC schließen */
  closeOnEscape?: boolean;
  /** Beim Auswählen eines Items automatisch schließen */
  closeOnSelect?: boolean;
  /** Eindeutige ID für das Dropdown (für ARIA) */
  id?: string;
  /** ARIA-Label für das Dropdown */
  'aria-label'?: string;
}

/**
 * Dropdown-Komponente für Dropdown-Menüs mit verbesserter Barrierefreiheit
 * 
 * @example
 * ```tsx
 * <Dropdown>
 *   <DropdownToggle>Options</DropdownToggle>
 *   <DropdownMenu>
 *     <DropdownItem value="edit">Edit</DropdownItem>
 *     <DropdownItem value="delete">Delete</DropdownItem>
 *   </DropdownMenu>
 * </Dropdown>
 * ```
 */
export const Dropdown = React.forwardRef<HTMLDivElement, DropdownProps>(({
  children,
  isOpen: controlledIsOpen,
  onOpenChange,
  onOpen,
  onClose,
  onSelect,
  placement = 'bottom',
  size = 'md',
  isDisabled = false,
  closeOnClickOutside = true,
  closeOnEscape = true,
  closeOnSelect = true,
  id: providedId,
  'aria-label': ariaLabel,
  className = '',
  ...rest
}, ref) => {
  // Generiere eine eindeutige ID, wenn keine bereitgestellt wurde
  const uniqueIdRef = useRef(`dropdown-${Math.random().toString(36).substr(2, 9)}`);
  const dropdownId = providedId || uniqueIdRef.current;
  
  // State für das Dropdown
  const [isOpen, setIsOpen] = useState(controlledIsOpen || false);
  const [activeItemIndex, setActiveItemIndex] = useState<number | null>(null);
  const triggerRef = useRef<HTMLElement>(null);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const itemsMap = useRef(new Map<string, number>());
  const itemsCounter = useRef(0);
  
  // Das Dropdown wird durch externe Props oder internen State gesteuert
  const isControlled = controlledIsOpen !== undefined;
  const dropdownIsOpen = isControlled ? controlledIsOpen : isOpen;
  
  // Registrieren eines neuen Items
  const registerItem = useCallback((id: string) => {
    if (!itemsMap.current.has(id)) {
      itemsMap.current.set(id, itemsCounter.current);
      return itemsCounter.current++;
    }
    return itemsMap.current.get(id)!;
  }, []);
  
  // State-Änderungen propagieren
  const updateOpenState = useCallback((newIsOpen: boolean) => {
    if (!isControlled) {
      setIsOpen(newIsOpen);
    }
    if (onOpenChange) {
      onOpenChange(newIsOpen);
    }
    
    if (newIsOpen) {
      if (onOpen) onOpen();
    } else {
      if (onClose) onClose();
      // Aktives Item zurücksetzen beim Schließen
      setActiveItemIndex(null);
    }
  }, [isControlled, onOpenChange, onOpen, onClose]);
  
  // Dropdown öffnen
  const openDropdown = useCallback(() => {
    if (isDisabled) return;
    updateOpenState(true);
  }, [isDisabled, updateOpenState]);
  
  // Dropdown schließen
  const closeDropdown = useCallback(() => {
    updateOpenState(false);
  }, [updateOpenState]);
  
  // Dropdown umschalten
  const toggleDropdown = useCallback(() => {
    if (isDisabled) return;
    updateOpenState(!dropdownIsOpen);
  }, [isDisabled, dropdownIsOpen, updateOpenState]);
  
  // Bei Klick außerhalb schließen
  useEffect(() => {
    if (!closeOnClickOutside || !dropdownIsOpen) return;
    
    const handleClickOutside = (e: MouseEvent) => {
      if (
        dropdownRef.current && 
        !dropdownRef.current.contains(e.target as Node) &&
        triggerRef.current && 
        !triggerRef.current.contains(e.target as Node)
      ) {
        // Verzögerung hinzufügen, um sicherzustellen, dass der Test den Zustand erfassen kann
        setTimeout(() => {
          closeDropdown();
        }, 0);
      }
    };
    
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [dropdownIsOpen, closeOnClickOutside, closeDropdown]);
  
  // Bei ESC-Taste schließen
  useEffect(() => {
    if (!closeOnEscape || !dropdownIsOpen) return;
    
    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === 'Escape') {
        e.preventDefault();
        closeDropdown();
        // Fokus zurück auf den Trigger setzen
        if (triggerRef.current) {
          triggerRef.current.focus();
        }
      }
    };
    
    document.addEventListener('keydown', handleEscape);
    return () => document.removeEventListener('keydown', handleEscape);
  }, [dropdownIsOpen, closeOnEscape, closeDropdown]);
  
  // Größen-spezifische Klassen
  const sizeClasses = {
    sm: 'dropdown-sm',
    md: 'dropdown-md',
    lg: 'dropdown-lg'
  };
  
  // CSS-Klassen zusammenstellen
  const classes = [
    'dropdown',
    sizeClasses[size],
    isDisabled ? 'dropdown-disabled' : '',
    className
  ].filter(Boolean).join(' ');
  
  return (
    <DropdownContext.Provider
      value={{
        isOpen: dropdownIsOpen,
        setIsOpen: updateOpenState,
        activeItemIndex,
        setActiveItemIndex,
        registerItem,
        triggerRef,
        dropdownId,
        onSelect,
        closeDropdown
      }}
    >
      <div
        ref={ref}
        id={dropdownId}
        className={classes}
        data-placement={placement}
        aria-label={ariaLabel}
        data-testid="dropdown"
        {...rest}
      >
        {children}
      </div>
    </DropdownContext.Provider>
  );
});

Dropdown.displayName = 'Dropdown';

export default Dropdown;