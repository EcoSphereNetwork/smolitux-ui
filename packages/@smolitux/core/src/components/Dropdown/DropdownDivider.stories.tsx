import type { Meta, StoryObj } from '@storybook/react';
import { DropdownDivider } from './DropdownDivider';

const meta: Meta<typeof DropdownDivider> = {
  title: 'Components/DropdownDivider',
  component: DropdownDivider,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: 'DropdownDivider',
  },
};

export const CustomStyle: Story = {
  args: {
    children: 'Custom DropdownDivider',
    className: 'custom-style',
  },
};

export const Interactive: Story = {
  args: {
    children: 'Interactive DropdownDivider',
    onClick: () => alert('Clicked!'),
  },
};
