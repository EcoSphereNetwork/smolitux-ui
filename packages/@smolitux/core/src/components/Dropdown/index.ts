// packages/@smolitux/core/src/components/Dropdown/index.ts
// Standard-Komponenten
export { default as Dropdown } from './Dropdown';
export { default as DropdownToggle } from './DropdownToggle';
export { default as DropdownMenu } from './DropdownMenu';
export { default as DropdownItem } from './DropdownItem';
export { default as DropdownDivider } from './DropdownDivider';

// Barrierefreie Komponenten
export { default as DropdownA11y } from './Dropdown.a11y';
export { default as DropdownToggleA11y } from './DropdownToggleA11y';
export { default as DropdownMenuA11y } from './DropdownMenuA11y';
export { default as DropdownItemA11y } from './DropdownItemA11y';
export { default as DropdownDividerA11y } from './DropdownDividerA11y';

// Fur Abwartskompatibilitat und bessere Lesbarkeit
export { DropdownA11y as DropdownAccessible } from './Dropdown.a11y';
export { DropdownToggleA11y as DropdownToggleAccessible } from './DropdownToggleA11y';
export { DropdownMenuA11y as DropdownMenuAccessible } from './DropdownMenuA11y';
export { DropdownItemA11y as DropdownItemAccessible } from './DropdownItemA11y';
export { DropdownDividerA11y as DropdownDividerAccessible } from './DropdownDividerA11y';

// Standard-Typen
export type { DropdownProps } from './Dropdown';
export type { DropdownToggleProps } from './DropdownToggle';
export type { DropdownMenuProps } from './DropdownMenu';
export type { DropdownItemProps } from './DropdownItem';
export type { DropdownDividerProps } from './DropdownDivider';

// Barrierefreie Typen
export type { DropdownA11yProps } from './Dropdown.a11y';
export type { DropdownToggleA11yProps } from './DropdownToggleA11y';
export type { DropdownMenuA11yProps } from './DropdownMenuA11y';
export type { DropdownItemA11yProps } from './DropdownItemA11y';
export type { DropdownDividerA11yProps } from './DropdownDividerA11y';