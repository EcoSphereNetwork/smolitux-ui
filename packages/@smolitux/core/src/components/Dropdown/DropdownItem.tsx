// packages/@smolitux/core/src/components/Dropdown/DropdownItem.tsx
import React, { useRef, useEffect } from 'react';
import { useDropdownContext } from './Dropdown';

export interface DropdownItemProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Inhalt des Items */
  children: React.ReactNode;
  /** Wert des Items (für onSelect) */
  value: string;
  /** Icon links vom Label */
  icon?: React.ReactNode;
  /** Icon rechts vom Label */
  rightIcon?: React.ReactNode;
  /** Ist das Item deaktiviert? */
  isDisabled?: boolean;
  /** Ist das Item aktiv? */
  isActive?: boolean;
  /** Callback beim Klick */
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
  /** Link, falls das Item ein Link ist */
  href?: string;
  /** Tastenkombination für das Item */
  shortcut?: string;
  /** Badge für das Item */
  badge?: React.ReactNode;
  /** Beschreibung für Screenreader */
  description?: string;
}

/**
 * DropdownItem-Komponente für Einträge in einem Dropdown
 * 
 * @example
 * ```tsx
 * <DropdownItem value="edit" icon={<EditIcon />}>Edit</DropdownItem>
 * ```
 */
export const DropdownItem = React.forwardRef<HTMLDivElement, DropdownItemProps>(({
  children,
  value,
  icon,
  rightIcon,
  isDisabled = false,
  isActive = false,
  onClick,
  href,
  shortcut,
  badge,
  description,
  className = '',
  ...rest
}, ref) => {
  const {
    activeItemIndex,
    registerItem,
    onSelect,
    closeDropdown
  } = useDropdownContext();
  
  const itemRef = useRef<HTMLDivElement | null>(null);
  const itemIndex = registerItem(value);
  
  // Prüfen, ob dieses Item aktiv ist
  const isActiveItem = activeItemIndex === itemIndex || isActive;
  
  // Ref-Forwarding
  useEffect(() => {
    if (itemRef.current && ref) {
      if (typeof ref === 'function') {
        ref(itemRef.current);
      } else {
        (ref as React.MutableRefObject<HTMLDivElement>).current = itemRef.current;
      }
    }
  }, [ref]);
  
  // Click-Handler für das Item
  const handleClick = (event: React.MouseEvent<HTMLDivElement>) => {
    if (isDisabled) {
      event.preventDefault();
      return;
    }
    
    // Custom onClick-Handler aufrufen
    if (onClick) {
      onClick(event);
    }
    
    // onSelect-Handler aufrufen
    if (onSelect) {
      onSelect(value);
    }
    
    // Dropdown schließen
    closeDropdown();
  };
  
  // Tastatur-Handler
  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    if (isDisabled) return;
    
    switch (e.key) {
      case 'Enter':
      case ' ':
        e.preventDefault();
        handleClick(e as unknown as React.MouseEvent<HTMLDivElement>);
        break;
      default:
        break;
    }
  };
  
  // CSS-Klassen zusammenstellen
  const classes = [
    // Basis-Klassen
    'flex items-center px-4 py-2 text-left w-full',
    'transition-colors duration-150',
    
    // Status-Klassen
    isDisabled ? 'opacity-50 cursor-not-allowed' : 'cursor-pointer',
    isActiveItem ? 'bg-primary-100 text-primary-800 dark:bg-primary-900/30 dark:text-primary-300' : 'hover:bg-gray-100 dark:hover:bg-gray-700',
    
    // Benutzerdefinierte Klassen
    className
  ].filter(Boolean).join(' ');
  
  // Basis-Element definieren (Link oder div)
  const ItemComponent: any = href && !isDisabled ? 'a' : 'div';
  const itemProps = href && !isDisabled ? { href } : {};
  
  return (
    <ItemComponent
      ref={itemRef}
      role="menuitem"
      tabIndex={isDisabled ? -1 : 0}
      aria-disabled={isDisabled ? 'true' : 'false'}
      aria-selected={isActiveItem}
      className={classes}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      data-value={value}
      aria-description={description}
      {...itemProps}
      {...rest}
    >
      {/* Icon (links) */}
      {icon && <span className="mr-2 flex-shrink-0" aria-hidden="true">{icon}</span>}
      
      {/* Hauptinhalt */}
      <span className="flex-grow truncate">
        {children}
        {description && <span className="sr-only">, {description}</span>}
      </span>
      
      {/* Shortcut */}
      {shortcut && (
        <span className="ml-4 text-xs text-gray-400 dark:text-gray-500" aria-label={`Tastenkombination: ${shortcut}`}>
          {shortcut}
        </span>
      )}
      
      {/* Badge */}
      {badge && <span className="ml-2">{badge}</span>}
      
      {/* Rechtes Icon */}
      {rightIcon && <span className="ml-2" aria-hidden="true">{rightIcon}</span>}
    </ItemComponent>
  );
});

DropdownItem.displayName = 'DropdownItem';

export default DropdownItem;