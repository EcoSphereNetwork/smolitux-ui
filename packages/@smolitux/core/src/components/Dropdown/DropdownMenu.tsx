// packages/@smolitux/core/src/components/Dropdown/DropdownMenu.tsx
import React, { useRef, useEffect } from 'react';
import { useDropdownContext } from './Dropdown';

export interface DropdownMenuProps extends React.HTMLAttributes<HTMLDivElement> {
  /** Kinder-Elemente (DropdownItems) */
  children: React.ReactNode;
  /** Maximale Höhe des Menüs (in px) */
  maxHeight?: number;
  /** Minimale Breite des Menüs (in px) */
  minWidth?: number;
  /** Maximale Breite des Menüs (in px) */
  maxWidth?: number;
  /** Variante des Menüs */
  variant?: 'default' | 'primary' | 'secondary';
  /** Größe des Menüs */
  size?: 'sm' | 'md' | 'lg';
  /** ARIA-Label für das Menü */
  'aria-label'?: string;
}

/**
 * DropdownMenu-Komponente für den Inhalt eines Dropdowns
 * 
 * @example
 * ```tsx
 * <DropdownMenu>
 *   <DropdownItem value="edit">Edit</DropdownItem>
 *   <DropdownItem value="delete">Delete</DropdownItem>
 * </DropdownMenu>
 * ```
 */
export const DropdownMenu = React.forwardRef<HTMLDivElement, DropdownMenuProps>(({
  children,
  maxHeight,
  minWidth,
  maxWidth,
  variant = 'default',
  size = 'md',
  'aria-label': ariaLabel,
  className = '',
  ...rest
}, ref) => {
  const {
    isOpen,
    dropdownId,
    activeItemIndex,
    setActiveItemIndex,
    triggerRef
  } = useDropdownContext();
  
  const menuRef = useRef<HTMLDivElement>(null);
  const itemsRef = useRef<(HTMLElement | null)[]>([]);
  
  // Varianten-spezifische Klassen
  const variantClasses = {
    default: 'bg-white dark:bg-gray-800 text-gray-800 dark:text-gray-200',
    primary: 'bg-primary-50 dark:bg-primary-900/20 text-primary-800 dark:text-primary-300',
    secondary: 'bg-secondary-50 dark:bg-secondary-900/20 text-secondary-800 dark:text-secondary-300'
  };
  
  // Größen-spezifische Klassen
  const sizeClasses = {
    sm: 'text-xs',
    md: 'text-sm',
    lg: 'text-base'
  };
  
  // CSS-Klassen zusammenstellen
  const classes = [
    // Basis-Klassen
    'rounded-md shadow-lg ring-1 ring-black ring-opacity-5 overflow-hidden',
    'absolute z-10 mt-1',
    
    // Variante
    variantClasses[variant],
    
    // Größe
    sizeClasses[size],
    
    // Benutzerdefinierte Klassen
    className
  ].filter(Boolean).join(' ');
  
  // Inline-Styles für Dimensionen
  const style: React.CSSProperties = {
    ...rest.style,
  };
  
  if (maxHeight) style.maxHeight = maxHeight;
  if (minWidth) style.minWidth = minWidth;
  if (maxWidth) style.maxWidth = maxWidth;
  
  // Tastatur-Navigation
  useEffect(() => {
    if (!isOpen || !menuRef.current) return;
    
    const handleKeyDown = (e: KeyboardEvent) => {
      if (!itemsRef.current.length) return;
      
      switch (e.key) {
        case 'ArrowDown':
        case 'Down': // Für ältere Browser
          e.preventDefault();
          setActiveItemIndex(prev => {
            const nextIndex = prev === null ? 0 : (prev + 1) % itemsRef.current.length;
            const nextItem = itemsRef.current[nextIndex];
            if (nextItem) nextItem.focus();
            return nextIndex;
          });
          break;
        case 'ArrowUp':
        case 'Up': // Für ältere Browser
          e.preventDefault();
          setActiveItemIndex(prev => {
            const nextIndex = prev === null ? itemsRef.current.length - 1 : (prev - 1 + itemsRef.current.length) % itemsRef.current.length;
            const nextItem = itemsRef.current[nextIndex];
            if (nextItem) nextItem.focus();
            return nextIndex;
          });
          break;
        case 'Home':
          e.preventDefault();
          if (itemsRef.current[0]) {
            itemsRef.current[0].focus();
            setActiveItemIndex(0);
          }
          break;
        case 'End':
          e.preventDefault();
          const lastIndex = itemsRef.current.length - 1;
          if (itemsRef.current[lastIndex]) {
            itemsRef.current[lastIndex].focus();
            setActiveItemIndex(lastIndex);
          }
          break;
        case 'Tab':
          // Wenn Tab gedrückt wird, schließen wir das Dropdown nicht,
          // sondern lassen die normale Tab-Navigation zu
          break;
        default:
          // Buchstaben-Navigation
          if (e.key.length === 1 && /[a-z0-9]/i.test(e.key)) {
            const char = e.key.toLowerCase();
            // Suche nach einem Item, das mit diesem Buchstaben beginnt
            const items = Array.from(menuRef.current?.querySelectorAll('[role="menuitem"]') || []);
            const index = items.findIndex((item, idx) => {
              // Beginne die Suche nach dem aktuellen Index
              const startIdx = activeItemIndex !== null ? (activeItemIndex + 1) % items.length : 0;
              const adjustedIdx = (startIdx + idx) % items.length;
              const text = item.textContent?.toLowerCase() || '';
              return text.startsWith(char);
            });
            
            if (index !== -1) {
              const adjustedIndex = activeItemIndex !== null ? (activeItemIndex + 1 + index) % items.length : index;
              const item = itemsRef.current[adjustedIndex];
              if (item) {
                item.focus();
                setActiveItemIndex(adjustedIndex);
              }
            }
          }
          break;
      }
    };
    
    menuRef.current.addEventListener('keydown', handleKeyDown);
    return () => {
      menuRef.current?.removeEventListener('keydown', handleKeyDown);
    };
  }, [isOpen, activeItemIndex, setActiveItemIndex]);
  
  // Fokus auf das erste Item setzen, wenn das Menü geöffnet wird
  useEffect(() => {
    if (isOpen && activeItemIndex === 0 && itemsRef.current[0]) {
      // Kurze Verzögerung, um sicherzustellen, dass das DOM aktualisiert wurde
      setTimeout(() => {
        itemsRef.current[0]?.focus();
      }, 0);
    }
  }, [isOpen, activeItemIndex]);
  
  // Kinder-Elemente mit Refs versehen
  const childrenWithRefs = React.Children.map(children, (child, index) => {
    if (!React.isValidElement(child)) return child;
    
    return React.cloneElement(child, {
      ref: (el: HTMLElement | null) => {
        itemsRef.current[index] = el;
        
        // Original-Ref weiterleiten
        const originalRef = (child as any).ref;
        if (originalRef) {
          if (typeof originalRef === 'function') {
            originalRef(el);
          } else if (originalRef.hasOwnProperty('current')) {
            originalRef.current = el;
          }
        }
      }
    });
  });
  
  if (!isOpen) return null;
  
  return (
    <div
      ref={menuRef}
      id={`${dropdownId}-menu`}
      className={classes}
      style={style}
      role="menu"
      aria-orientation="vertical"
      aria-labelledby={`${dropdownId}-toggle`}
      aria-label={ariaLabel}
      tabIndex={-1}
      {...rest}
    >
      {childrenWithRefs}
    </div>
  );
});

DropdownMenu.displayName = 'DropdownMenu';

export default DropdownMenu;