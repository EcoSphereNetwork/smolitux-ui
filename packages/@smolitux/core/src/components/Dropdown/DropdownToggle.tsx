// packages/@smolitux/core/src/components/Dropdown/DropdownToggle.tsx
import React, { useRef, useEffect } from 'react';
import { useDropdownContext } from './Dropdown';

export interface DropdownToggleProps extends React.HTMLAttributes<HTMLButtonElement> {
  /** Inhalt des Toggle-Buttons */
  children: React.ReactNode;
  /** Icon links vom Label */
  icon?: React.ReactNode;
  /** Icon rechts vom Label */
  rightIcon?: React.ReactNode;
  /** Variante des Buttons */
  variant?: 'default' | 'primary' | 'secondary' | 'outline' | 'ghost';
  /** Größe des Buttons */
  size?: 'sm' | 'md' | 'lg';
  /** Ist der Button deaktiviert? */
  isDisabled?: boolean;
  /** Volle Breite */
  fullWidth?: boolean;
}

/**
 * DropdownToggle-Komponente für den Trigger eines Dropdowns
 * 
 * @example
 * ```tsx
 * <DropdownToggle>Options</DropdownToggle>
 * ```
 */
export const DropdownToggle = React.forwardRef<HTMLButtonElement, DropdownToggleProps>(({
  children,
  icon,
  rightIcon,
  variant = 'default',
  size = 'md',
  isDisabled = false,
  fullWidth = false,
  className = '',
  onClick,
  ...rest
}, ref) => {
  const {
    isOpen,
    setIsOpen,
    triggerRef,
    dropdownId,
    setActiveItemIndex
  } = useDropdownContext();
  
  const buttonRef = useRef<HTMLButtonElement | null>(null);
  
  // Ref-Forwarding
  useEffect(() => {
    if (buttonRef.current) {
      // Ref an den übergeordneten Context weitergeben
      if (triggerRef) {
        (triggerRef as React.MutableRefObject<HTMLElement>).current = buttonRef.current;
      }
      
      // Ref an den Consumer weitergeben
      if (ref) {
        if (typeof ref === 'function') {
          ref(buttonRef.current);
        } else {
          (ref as React.MutableRefObject<HTMLButtonElement>).current = buttonRef.current;
        }
      }
    }
  }, [ref, triggerRef]);
  
  // Varianten-spezifische Klassen
  const variantClasses = {
    default: 'bg-white text-gray-800 border border-gray-300 hover:bg-gray-50 dark:bg-gray-800 dark:text-gray-200 dark:border-gray-700 dark:hover:bg-gray-700',
    primary: 'bg-primary-600 text-white border border-primary-600 hover:bg-primary-700 dark:bg-primary-700 dark:border-primary-700 dark:hover:bg-primary-800',
    secondary: 'bg-secondary-600 text-white border border-secondary-600 hover:bg-secondary-700 dark:bg-secondary-700 dark:border-secondary-700 dark:hover:bg-secondary-800',
    outline: 'bg-transparent text-gray-800 border border-gray-300 hover:bg-gray-50 dark:text-gray-200 dark:border-gray-700 dark:hover:bg-gray-800',
    ghost: 'bg-transparent text-gray-800 border-none hover:bg-gray-100 dark:text-gray-200 dark:hover:bg-gray-800'
  };
  
  // Größen-spezifische Klassen
  const sizeClasses = {
    sm: 'text-xs py-1 px-2',
    md: 'text-sm py-2 px-3',
    lg: 'text-base py-2.5 px-4'
  };
  
  // CSS-Klassen zusammenstellen
  const classes = [
    // Basis-Klassen
    'inline-flex items-center justify-center rounded-md font-medium transition-colors',
    'focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary-500',
    
    // Variante
    variantClasses[variant],
    
    // Größe
    sizeClasses[size],
    
    // Status
    isDisabled ? 'opacity-50 cursor-not-allowed' : 'cursor-pointer',
    
    // Weitere Optionen
    fullWidth ? 'w-full' : '',
    
    // Benutzerdefinierte Klassen
    className
  ].filter(Boolean).join(' ');
  
  // Click-Handler
  const handleClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (isDisabled) {
      e.preventDefault();
      return;
    }
    
    setIsOpen(!isOpen);
    
    // Wenn das Dropdown geöffnet wird, das erste Item aktivieren
    if (!isOpen) {
      setActiveItemIndex(0);
    }
    
    // Custom onClick-Handler aufrufen
    if (onClick) {
      onClick(e);
    }
  };
  
  // Tastatur-Handler
  const handleKeyDown = (e: React.KeyboardEvent<HTMLButtonElement>) => {
    if (isDisabled) return;
    
    switch (e.key) {
      case 'ArrowDown':
      case 'Down': // Für ältere Browser
        e.preventDefault();
        setIsOpen(true);
        setActiveItemIndex(0);
        break;
      case 'ArrowUp':
      case 'Up': // Für ältere Browser
        e.preventDefault();
        setIsOpen(true);
        setActiveItemIndex(-1); // Das letzte Item aktivieren
        break;
      case 'Enter':
      case ' ':
        e.preventDefault();
        setIsOpen(!isOpen);
        break;
      default:
        break;
    }
  };
  
  return (
    <button
      ref={buttonRef}
      type="button"
      className={classes}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      aria-haspopup="menu"
      aria-expanded={isOpen}
      aria-controls={isOpen ? `${dropdownId}-menu` : undefined}
      aria-disabled={isDisabled ? 'true' : 'false'}
      disabled={isDisabled}
      {...rest}
    >
      {/* Icon (links) */}
      {icon && <span className="mr-2">{icon}</span>}
      
      {/* Hauptinhalt */}
      <span>{children}</span>
      
      {/* Icon (rechts) oder Standard-Dropdown-Icon */}
      {rightIcon ? (
        <span className="ml-2">{rightIcon}</span>
      ) : (
        <svg
          className="ml-2 h-4 w-4"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          aria-hidden="true"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            d="M19 9l-7 7-7-7"
          />
        </svg>
      )}
    </button>
  );
});

DropdownToggle.displayName = 'DropdownToggle';

export default DropdownToggle;