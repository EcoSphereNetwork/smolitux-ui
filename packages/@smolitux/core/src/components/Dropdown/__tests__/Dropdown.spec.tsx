import React from 'react';
import renderer from 'react-test-renderer';
import { Dropdown } from '../Dropdown';
import { DropdownToggle } from '../DropdownToggle';
import { DropdownMenu } from '../DropdownMenu';
import { DropdownItem } from '../DropdownItem';

describe('Dropdown Snapshots', () => {
  test('renders closed dropdown', () => {
    const tree = renderer
      .create(
        <Dropdown id="dropdown-test">
          <DropdownToggle>Menu</DropdownToggle>
          <DropdownMenu>
            <DropdownItem value="profile">Profile</DropdownItem>
          </DropdownMenu>
        </Dropdown>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders open dropdown', () => {
    const tree = renderer
      .create(
        <Dropdown id="dropdown-test" isOpen>
          <DropdownToggle>Menu</DropdownToggle>
          <DropdownMenu>
            <DropdownItem value="profile">Profile</DropdownItem>
          </DropdownMenu>
        </Dropdown>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
