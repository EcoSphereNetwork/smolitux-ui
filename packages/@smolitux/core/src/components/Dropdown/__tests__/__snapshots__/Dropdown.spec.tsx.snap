// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Snapshots renders closed dropdown 1`] = `
<div
  className="dropdown dropdown-md"
  data-placement="bottom"
  data-testid="dropdown"
  id="dropdown-test"
>
  <button
    aria-disabled="false"
    aria-expanded={false}
    aria-haspopup="menu"
    className="inline-flex items-center justify-center rounded-md font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary-500 bg-white text-gray-800 border border-gray-300 hover:bg-gray-50 dark:bg-gray-800 dark:text-gray-200 dark:border-gray-700 dark:hover:bg-gray-700 text-sm py-2 px-3 cursor-pointer"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span>
      Menu
    </span>
    <svg
      aria-hidden="true"
      className="ml-2 h-4 w-4"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 9l-7 7-7-7"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
      />
    </svg>
  </button>
</div>
`;

exports[`Dropdown Snapshots renders open dropdown 1`] = `
<div
  className="dropdown dropdown-md"
  data-placement="bottom"
  data-testid="dropdown"
  id="dropdown-test"
>
  <button
    aria-controls="dropdown-test-menu"
    aria-disabled="false"
    aria-expanded={true}
    aria-haspopup="menu"
    className="inline-flex items-center justify-center rounded-md font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary-500 bg-white text-gray-800 border border-gray-300 hover:bg-gray-50 dark:bg-gray-800 dark:text-gray-200 dark:border-gray-700 dark:hover:bg-gray-700 text-sm py-2 px-3 cursor-pointer"
    disabled={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span>
      Menu
    </span>
    <svg
      aria-hidden="true"
      className="ml-2 h-4 w-4"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 9l-7 7-7-7"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
      />
    </svg>
  </button>
  <div
    aria-labelledby="dropdown-test-toggle"
    aria-orientation="vertical"
    className="rounded-md shadow-lg ring-1 ring-black ring-opacity-5 overflow-hidden absolute z-10 mt-1 bg-white dark:bg-gray-800 text-gray-800 dark:text-gray-200 text-sm"
    id="dropdown-test-menu"
    role="menu"
    style={{}}
    tabIndex={-1}
  >
    <div
      aria-disabled="false"
      aria-selected={false}
      className="flex items-center px-4 py-2 text-left w-full transition-colors duration-150 cursor-pointer hover:bg-gray-100 dark:hover:bg-gray-700"
      data-value="profile"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="menuitem"
      tabIndex={0}
    >
      <span
        className="flex-grow truncate"
      >
        Profile
      </span>
    </div>
  </div>
</div>
`;
